### ISSUE TITLE
Error 400

### ISSUE THREAD

**Issue by richiefrich** [terraform-google-modules/terraform-google-kubernetes-engine#2222]
### TL;DR

On the initial run of building a GKE cluster it works. then if I rerun the module, it updated the node-pools and errors with "Error 400: At least one of"

### Expected behavior

Expect it to just update the node-pool

### Observed behavior

module.primary-cluster.google_container_cluster.primary: Modifications complete after 33m56s [id=projects/ss-pp-fleet-test01-p/locations/us-central1/clusters/services-01]
module.primary-cluster.google_container_node_pool.pools["services-node-pool"]: Modifying... [id=projects/ss-pp-fleet-test01-p/locations/us-central1/clusters/services-01/nodePools/services-node-pool-73f9]

Error: googleapi: Error 400: At least one of ['node_version', 'image_type', 'updated_node_pool', 'locations', 'workload_metadata_config', 'upgrade_settings', 'kubelet_config', 'linux_node_config', 'tags', 'taints', 'labels', 'node_network_config', 'gcfs_config', 'gvnic', 'confidential_nodes', 'logging_config', 'fast_socket', 'resource_labels', 'accelerators', 'windows_node_config', 'machine_type', 'disk_type', 'disk_size_gb', 'storage_pools', 'containerd_config', 'resource_manager_tags', 'performance_monitoring_unit', 'queued_provisioning', 'max_run_duration'] must be specified.
Details:
[
  {
    "@type": "type.googleapis.com/google.rpc.RequestInfo",
    "requestId": "0x1d7a40eb9326f795"
  }
]
, badRequest

### Terraform Configuration

```hcl
Using Terragrunt but here is my node-pool config

  node_pools = [
    {
      name                             = "services-node-pool"
      autoscaling                      = true
      auto_repair                      = true
      auto_upgrade                     = true
      min_count                        = 1
      max_count                        = 4
      node_count                       = 1
      image_type                       = "COS_CONTAINERD"
      machine_type                     = "n2-standard-8"
      enable_secure_boot               = true
      service_account                  = "services-gke-sa@${local.project_id}.iam.gserviceaccount.com"
      node_pools_create_before_destroy = true
    },
  ]
```

### Terraform Version

```sh
$ terraform  -version
Terraform v1.5.7
on linux_amd64

This used to work, I know its an older version of Terraform.
```

**Comment by hellolin324** [comment]
I am having the same issue, it tries to delete the kubelet_config under the node_config block everytime I run a tf apply, and throws me this error:
googleapi: Error 400: At least one of ['node_version', 'image_type', 'updated_node_pool', 'locations', 'workload_metadata_config', 'upgrade_settings', 'kubelet_config', 'linux_node_config', 'tags', 'taints', 'labels', 'node_network_config', 'gcfs_config', 'gvnic', 'confidential_nodes', 'logging_config', 'fast_socket', 'resource_labels', 'accelerators', 'windows_node_config', 'machine_type', 'disk_type', 'disk_size_gb', 'storage_pools', 'containerd_config', 'resource_manager_tags', 'performance_monitoring_unit', 'queued_provisioning', 'max_run_duration'] must be specified.
│ Details:
│ [
│   {
│     "@type": "type.googleapis.com/google.rpc.RequestInfo",
│     "requestId": "0x1d114922d7dd5a2f"
│   }
│ ]
│ , badRequest

It does not affect any functionality as the gke cluster still gets to be created but I can't make the error go away.

**Comment by richiefrich** [comment]
I would still like to know what changed, as even using older versions have worked up until maybe 2 weeks ago?
Has Google updated the older versions of the resources?

**Comment by hellolin324** [comment]
I have no idea what had changed, but before it was throwing me the same error on a different tag that belonged in another map, and I was not able to fit it into a list of objects due to tf's constraints. Updating the Google provider worked for me as that allowed me to use the latest GKE module here, and that fixed it. Tested with using different flags and they all worked fine, revert back to old provider and error was back.

**Comment by rgddos** [comment]
> I would still like to know what changed, as even using older versions have worked up until maybe 2 weeks ago? Has Google updated the older versions of the resources?


We encountered same issue and our CI pipelines were continuously failing. This started about two weeks ago, although everything had been working fine before that. We were using an older provider (3.x.x), which made it challenging to upgrade the provider



To address this, we implemented the following hardcoded configuration in our module:

```hcl
kubelet_config {
  cpu_manager_policy   = "static"
  cpu_cfs_quota        = false
}

**Comment by hellolin324** [comment]
```
kubelet_config {
  cpu_manager_policy   = "static"
  cpu_cfs_quota        = false
}
```

This is what I tried but couldn't work, so instead I forked this repo and added an additional line in the `lifecycle{}` block of the node_pool resource in the cluster.tf file, which also worked with the older provider.

Can you share your entire block of tf code for the config please?

**Comment by KatrinaHoffert** [comment]
A drive by comment that this was probably what https://github.com/GoogleCloudPlatform/magic-modules/pull/11971 fixed. Looks like version 6.7.0 probably is the first that fixed it. In short, new server-side computed values are set and since the TF provider is too old to know about them, it thinks it should remove them and constructs an invalid request to do so (the request ends up looking as if it were updating nothing, hence the error).

I didn't test it, but suspect `ignore_changes` can be used as a temporary workaround when updating the provider isn't an option, as it would tell TF to ignore that the field has "changed" and thus prevent it from making the invalid request to update it.

**Comment by heldersepu** [comment]
> A drive by comment that this was probably what [GoogleCloudPlatform/magic-modules#11971](https://github.com/GoogleCloudPlatform/magic-modules/pull/11971) fixed. Looks like version 6.7.0 probably is the first that fixed it...
> 
I'm getting this same error and I'm at v6.16

```
terraform init

Initializing the backend...
Initializing modules...

Initializing provider plugins...
- Reusing previous version of hashicorp/google-beta from the dependency lock file
- Reusing previous version of hashicorp/kubernetes from the dependency lock file
- Reusing previous version of hashicorp/random from the dependency lock file
- Reusing previous version of hashicorp/google from the dependency lock file
- Reusing previous version of hashicorp/helm from the dependency lock file
- Using previously-installed hashicorp/google-beta v4.85.0
- Using previously-installed hashicorp/kubernetes v2.35.1
- Using previously-installed hashicorp/random v3.6.3
- Using previously-installed hashicorp/google v6.16.0
- Using previously-installed hashicorp/helm v2.17.0
```

**Comment by heldersepu** [comment]
the version i was using was very old change to latest and all good:

```
module "eks_cluster" {
  source                     = "terraform-google-modules/kubernetes-engine/google//modules/beta-public-cluster"
  version                    = "~> 35.0"
```

**Comment by pramodsetlur** [comment]
> A drive by comment that this was probably what [GoogleCloudPlatform/magic-modules#11971](https://github.com/GoogleCloudPlatform/magic-modules/pull/11971) fixed. Looks like version 6.7.0 probably is the first that fixed it. In short, new server-side computed values are set and since the TF provider is too old to know about them, it thinks it should remove them and constructs an invalid request to do so (the request ends up looking as if it were updating nothing, hence the error).

I think @KatrinaHoffert's [response](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/2222#issuecomment-2576344471) sums it up. We are currently using `~> v26` (so [v26.1.1](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/v26.1.1/modules/private-cluster-update-variant/cluster.tf)) of the `private-cluster-update-variant` and this comes with a provider constraint of version [">= 4.51.0, < 5.0, !=4.65.0, !=4.65.1" ](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/139f98e5572b0e3501668ea1cec51cdd9438b5c4/modules/private-cluster-update-variant/versions.tf#L24) for the `hasicorp/google` provider. I am wondering if the changes made to the provider will be released to v4 (4.85.0 is the latest and the one we use) as well (similar to the patches: [6.7.0 and 5.44.2](https://github.com/hashicorp/terraform-provider-google/issues/19792#issuecomment-2412024000)). Sorry for the tag @wyardley - just curious if you would know. 

We are currently not in a position to upgrade to the latest version of the terraform module (v35 as mentioned above) since there are many breaking changes between v26 and v35.

**Comment by apeabody** [comment]
> > A drive by comment that this was probably what [GoogleCloudPlatform/magic-modules#11971](https://github.com/GoogleCloudPlatform/magic-modules/pull/11971) fixed. Looks like version 6.7.0 probably is the first that fixed it...
> 
> I'm getting this same error and I'm at v6.16
> 
> ```
> terraform init
> 
> Initializing the backend...
> Initializing modules...
> 
> Initializing provider plugins...
> - Reusing previous version of hashicorp/google-beta from the dependency lock file
> - Reusing previous version of hashicorp/kubernetes from the dependency lock file
> - Reusing previous version of hashicorp/random from the dependency lock file
> - Reusing previous version of hashicorp/google from the dependency lock file
> - Reusing previous version of hashicorp/helm from the dependency lock file
> - Using previously-installed hashicorp/google-beta v4.85.0
> - Using previously-installed hashicorp/kubernetes v2.35.1
> - Using previously-installed hashicorp/random v3.6.3
> - Using previously-installed hashicorp/google v6.16.0
> - Using previously-installed hashicorp/helm v2.17.0
> ```

Hi @heldersepu - Looks like your `google-beta` provider is still at `v4.85.0`, which is used by the beta cluster sub-modules.  `terraform providers` can likely help identify the source of the constraint, and then I recommend updating to the most recent version of `terraform-google-kubernetes-engine` and `hashicorp/google-beta`.

```
> - Using previously-installed hashicorp/google-beta v4.85.0
```

**Comment by heldersepu** [comment]
@apeabody thanks for the reply I did got it working 
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/2222#issuecomment-2591348987

**Comment by apeabody** [comment]
> A drive by comment that this was probably what [GoogleCloudPlatform/magic-modules#11971](https://github.com/GoogleCloudPlatform/magic-modules/pull/11971) fixed. Looks like version 6.7.0 probably is the first that fixed it. In short, new server-side computed values are set and since the TF provider is too old to know about them, it thinks it should remove them and constructs an invalid request to do so (the request ends up looking as if it were updating nothing, hence the error).
> 
> I didn't test it, but suspect `ignore_changes` can be used as a temporary workaround when updating the provider isn't an option, as it would tell TF to ignore that the field has "changed" and thus prevent it from making the invalid request to update it.

Agreed - In general it is highly recommended to use recent Terraform Google Provider (`hashicorp/google` and `hashicorp/google-beta`) and terraform-google-kubernetes-engine for best compatibility with the underlying GKE API.  While older versions will often work, this isn't guaranteed long-term.  For posterity, it appears the observed issues is resolved once upgrade to a recent Terraform Google Provider and terraform-google-kubernetes-engine version.

@richiefrich - The current minimum Terraform version for this module is v1.3, so your Terraform v1.5.7 version should work fine.

@pramodsetlur - With the release of [Terraform Google Provider v6](https://cloud.google.com/blog/products/management-tools/announcing-terraform-google-provider-6-0-0), unfortunately I am not expecting any further updates to Terraform Google Provider v4.
